### ISSUE TITLE
Deterministic indexing operation fails in indices size check / missing broadcast

### ISSUE THREAD

**Issue by ptrblck** [pytorch/pytorch#61032]
## 🐛 Bug

Reported in the [forum](https://discuss.pytorch.org/t/torch-deterministic-algorithms-error/125200) by [Nexkira](https://discuss.pytorch.org/u/Nexkira) (thanks for reporting this issue and providing the initial code snippet!).

## To Reproduce

```python
import torch
torch.use_deterministic_algorithms(True)

x = torch.arange(10).float().cuda()
x[x>2.] = 1.
print(x)
> RuntimeError: linearIndex.numel()*sliceSize*nElemBefore == value.numel()INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Indexing.cu":253, please report a bug to PyTorch. number of flattened indices did not match number of elements in the value tensor71
```

[Line of code](https://github.com/pytorch/pytorch/blob/10fc58620e58fb4cf6acd2af4a13f6d73dd0f41a/aten/src/ATen/native/cuda/Indexing.cu#L253) raising the issue.
Based on the error message a broadcast seems to be missing for the `value` (the shapes are given as 7 and 1).

## Environment

Reproduced using the nightly pip wheel with CUDA10.2 and CUDA11.1 on a P100 and 3090.

## Additional context

The code runs fine, when the default `torch.use_deterministic_algorithms(False)` setting or the CPU is used.

CC @eqy

**Comment by ngimel** [comment]
Preferred fix for this is #57515, it requires untangling stuff in `python_variable_indexing.cpp` though. I have a tentative PR for this, but "fast deterministic path" I enable isn't taken because by the time `index` function is called, `val` is copied to cuda by python_variable_indexing. 
The proposed PR #61065 doesn't enable full broadcasting, only for a 1-element tensor, and 1-element is always better to dispatch to masked_fill (added bonus is no synchronization, hence capturable).

**Comment by xwang233** [comment]
https://github.com/pytorch/pytorch/issues/39227 is likely related

**Comment by eqy** [comment]
> Preferred fix for this is #57515, it requires untangling stuff in `python_variable_indexing.cpp` though. I have a tentative PR for this, but "fast deterministic path" I enable isn't taken because by the time `index` function is called, `val` is copied to cuda by python_variable_indexing.
> The proposed PR #61065 doesn't enable full broadcasting, only for a 1-element tensor, and 1-element is always better to dispatch to masked_fill (added bonus is no synchronization, hence capturable).

Sure, `masked_select` is definitely a more efficient way to do this kind of operation. Are there any dependencies or PRs that need to be landed first before we move in that direction?

**Comment by ngimel** [comment]
I tried to enable dispatching to masked_fill https://github.com/pytorch/pytorch/compare/ngimel/masked_select?expand=1, and it works with c++ APIs, but with python `value.device().is_cpu()` check fails, because value is copied to cpu before dispatching to `index`, in `python_variable_indexing.cpp`. `python_variable_indexing` is pretty convoluted, and has to be somehow fixed to not copy values over-eagerly. There's probably a legitimate case where values should be copied, but it prevents masked_fill optimization, so all this has to be untangled, and I didn't do it. @eqy, if you can try doing it that would be great.

**Comment by eqy** [comment]
> I tried to enable dispatching to masked_fill https://github.com/pytorch/pytorch/compare/ngimel/masked_select?expand=1, and it works with c++ APIs, but with python `value.device().is_cpu()` check fails, because value is copied to cpu before dispatching to `index`, in `python_variable_indexing.cpp`. `python_variable_indexing` is pretty convoluted, and has to be somehow fixed to not copy values over-eagerly. There's probably a legitimate case where values should be copied, but it prevents masked_fill optimization, so all this has to be untangled, and I didn't do it. @eqy, if you can try doing it that would be great.

I'm not familiar with `python_variable_indexing` but I can take a look. At the moment I'm not even sure where the copy to CPU happens; at first glance I thought it was in `THPVariable_setitem` (https://github.com/pytorch/pytorch/blob/61f946bba612a34df25e6cf53cef03f7bfa8abaa/torch/csrc/autograd/python_variable_indexing.cpp#L372), but it looks like when I run this with a trivial CUDA tensor it is allocating the value tensor on GPU.
Is there a different minimal example I should be testing? I was checking
```
>>> import torch
>>> a = torch.randn(1000, device='cuda')
>>> a[a>0] = 100
```

**Comment by ngimel** [comment]
Right, your mimimum example reproduces the problem, and AFAIK it happens here https://github.com/pytorch/pytorch/blob/61f946bba612a34df25e6cf53cef03f7bfa8abaa/torch/csrc/autograd/python_variable_indexing.cpp#L374, `py_value` is copied to the device of `self_`. It is the right (or at least bc-compatible) thing to do for some cases, but it's an unneeded overhead for `x[mask]=scalar`

**Comment by eqy** [comment]
OK, I think I was misunderstanding the situation. So in the scalar case, we _do_ want the value tensor to remain on CPU.

**Comment by ngimel** [comment]
Yes, to be able to dispatch to `masked_fill_` (masked_fill accepts only cpu scalar).

**Comment by eqy** [comment]
I'm trying to see if I can get the conditions right to force dispatching to `masked_select` but it looks like I get a ```RuntimeError: expand(CUDABoolType{[100, 1]}, size=[100]): the number of sizes provided (1) must be greater or equal to the number of dimensions in the tensor (2)```.
Is the example I'm using working on your end?
```
>>> import torch
>>> a = torch.randn(1000, device='cuda')
>>> a[a>0] = 100
```

**Comment by ngimel** [comment]
I don't know, don't have a build with the patch. Looks like the wrong number of `unsqueeze_` calls?

**Comment by ngimel** [comment]
Fixed in #61612
